// @swimlane/ngx-graph/esm2015/lib/models/layout.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Layout() { }
if (false) {
    /** @type {?|undefined} */
    Layout.prototype.settings;
    /**
     * @param {?} graph
     * @return {?}
     */
    Layout.prototype.run = function (graph) { };
    /**
     * @param {?} graph
     * @param {?} edge
     * @return {?}
     */
    Layout.prototype.updateEdge = function (graph, edge) { };
    /**
     * @param {?} draggingNode
     * @param {?} $event
     * @return {?}
     */
    Layout.prototype.onDragStart = function (draggingNode, $event) { };
    /**
     * @param {?} draggingNode
     * @param {?} $event
     * @return {?}
     */
    Layout.prototype.onDrag = function (draggingNode, $event) { };
    /**
     * @param {?} draggingNode
     * @param {?} $event
     * @return {?}
     */
    Layout.prototype.onDragEnd = function (draggingNode, $event) { };
}
